LNHANCE inquisition (CSFS, INTERNALKEY)

reardencode opened the pull request:
This adds CHECKSIGFROMSTACK and INTERNALKEY to bitcoin inquisition for review and testing.

(Related to https://github.com/bitcoin/bitcoin/pull/29198)

<!--
*** Please remove the following help text before submitting: ***

Pull requests without a rationale and clear improvement may be closed
immediately.

GUI-related pull requests should be opened against
https://github.com/bitcoin-core/gui
first. See CONTRIBUTING.md
-->

<!--
Please provide clear motivation for your patch and explain how it improves
Bitcoin Core user experience or Bitcoin Core developer experience
significantly:

* Any test improvements or new tests that improve coverage are always welcome.
* All other changes should have accompanying unit tests (see `src/test/`) or
  functional tests (see `test/`). Contributors should note which tests cover
  modified code. If no tests exist for a region of modified code, new tests
  should accompany the change.
* Bug fixes are most welcome when they come with steps to reproduce or an
  explanation of the potential issue as well as reasoning for the way the bug
  was fixed.
* Features are welcome, but might be rejected due to design or scope issues.
  If a feature is based on a lot of dependencies, contributors should first
  consider building the system outside of Bitcoin Core, if possible.
* Refactoring changes are only accepted if they are required for a feature or
  bug fix or otherwise improve developer experience significantly. For example,
  most "code style" refactoring changes require a thorough explanation why they
  are useful, what downsides they have and why they *significantly* improve
  developer experience or avoid serious programming bugs. Note that code style
  is often a subjective matter. Unless they are explicitly mentioned to be
  preferred in the [developer notes](/doc/developer-notes.md), stylistic code
  changes are usually rejected.
-->

<!--
Bitcoin Core has a thorough review process and even the most trivial change
needs to pass a lot of eyes and requires non-zero or even substantial time
effort to review. There is a huge lack of active reviewers on the project, so
patches often sit for a long time.
-->

reardencode committed:
consensus: add BIN-2024-000[34] deployment params

Activates OP_CHECKSIGFROMSTACK, OP_CHECKSIGFROMSTACKVERIFY, and
OP_INTERNALKEY on regtest and signet.

always active on regtest
active from 2024-01-12 until 2034-01-12 on signet
never active on testnet, mainnet

reardencode committed:
Add OP_INTERNALKEY for Tapscript

Testing is minimal, but so is the code.

ajtowns commented on src/script/script.cpp:
```diff
@@ -160,7 +164,8 @@ unsigned int CScript::GetSigOpCount(bool fAccurate) const
             break;
-        if (opcode == OP_CHECKSIG || opcode == OP_CHECKSIGVERIFY)
+        if (opcode == OP_CHECKSIG || opcode == OP_CHECKSIGVERIFY ||
+            opcode == OP_CHECKSIGFROMSTACK || opcode == OP_CHECKSIGFROMSTACKVERIFY)
```
I don't think this change is correct: `GetSigOpCount` is for legacy, p2sh and p2wsh scripts, not taproot scripts.

reardencode replied:
Good catch. CHECKSIGFROMSTACKVERIFY belongs here, but not CHECKSIGFROMSTACK.

ajtowns commented on src/validation.cpp:
```diff
@@ -1007,11 +1007,14 @@ bool MemPoolAccept::PolicyScriptChecks(const ATMPArgs& args, Workspace& ws)
     const CTransaction& tx = *ws.m_ptx;
     TxValidationState& state = ws.m_state;
 
+    const bool lnhance_active = DeploymentActiveAfter(m_active_chainstate.m_chain.Tip(), m_active_chainstate.m_chainman, Consensus::DEPLOYMENT_LNHANCE);
```
My gut feel is that it would make more sense to have separate deployments for the separate BIPs here, even though they might be combined into a single deployment for mainnet.

reardencode replied:
Makes sense, especially for inquisition. I was torn about it.

reardencode replied:
I made it separate flags, but one deployment for the moment. If you want me to fully split the deployments for inquisition, lmk.

ajtowns commented on src/script/interpreter.cpp:
```diff
@@ -1963,6 +2075,9 @@ static bool ExecuteWitnessScript(const Span<const valtype>& stack_span, const CS
                 // Note how this condition would not be reached if an unknown OP_SUCCESSx was found
                 return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
             }
+            if ((flags & SCRIPT_VERIFY_LNHANCE) && (opcode == OP_CHECKSIGFROMSTACK || opcode == OP_INTERNALKEY)) {
```
Writing this as:

```c++
    if (opcode == CSFS || opcode == IKEY) {
        if (flags & DISCOURAGE_LNHANCE) {
            return set_error(serror, SCRIPT_ERR_DISCOURAGE_OP_SUCCESS);
            // discourage OP_SUCCESS behaviour -- applies even in unexecuted IF/ELSE branches
        }
        if (flags & LNHANCE) continue;
        return set_success(serror); // OP_SUCCESS behaviour
    }
    if (IsOpSuccess(opcode)) { ... }
```

would avoid the need for the special case in `TrimFlags`, I believe.

ajtowns reviewed (commented):
Deployment params should be the last commit not the first -- that way the implementation commits can more easily be cherry-picked, merged and tested first.

reardencode commented:
> Deployment params should be the last commit not the first -- that way the implementation commits can more easily be cherry-picked, merged and tested first.

IIUC it would have to be:
1. implementation, with cpp unit tests
2. deployment params
3. functional tests

Because functional tests won't be possible until regtest deployment. This makes sense on this PR, but less so on my core one (from whence this was ported) because there it would have involved splitting the CTV functional testing commits from the rest of the CTV changes. For CSFS and IKEY I only did cpp unit tests, which may or may not be sufficient (especially here where they're split from CTV's deployment testing).

reardencode commented:
Addressed all of your comments except fully splitting out the deployment. Happy to do that too if you think it makes sense with the other changes in place. Thanks much for the review!

ajtowns commented on src/kernel/chainparams.cpp:
```diff
@@ -130,6 +130,8 @@ class CMainParams : public CChainParams {
         consensus.vDeployments[Consensus::DEPLOYMENT_CHECKTEMPLATEVERIFY] = SetupDeployment{.bip = 119, .bip_version = 0, .never = true};
         consensus.vDeployments[Consensus::DEPLOYMENT_ANYPREVOUT] = SetupDeployment{.bip = 118, .bip_version = 0, .never = true};
 
+        consensus.vDeployments[Consensus::DEPLOYMENT_LNHANCE] = SetupDeployment{.bip = 999, .bip_version = 0, .never = true};
```
Can you submit the proposal specs via https://github.com/bitcoin-inquisition/binana/pulls so we can have some canonical numbers here? Hopefully converting to md isn't too much of a pain?

reardencode replied:
Added PRs for BIN-2024-000[34].

ajtowns replied:
Oh, these lines need to be updated after #46, swapping `.bip=X, .bip_version=Y` out for `.year=2024, .number=3, .revision=0` or so.

reardencode committed:
remove ECDSA verification from CSFSV

reardencode commented:
Updated for BINANA and to remove ECDSA verification from CHECKSIGFROMSTACKVERIFY on legacy/witnessv0 (BIN-2024-0003 rev1).

rot13maxi commented on src/kernel/chainparams.cpp:
```diff
@@ -394,6 +400,20 @@ class SigNetParams : public CChainParams {
+            .revision = 1,
+            .start = 1705046400, // 2024-01-12
+            .timeout = 2020665600, // 2034-01-12
+        };
```
I was under the impression that LNHANCE is a proposal to do CTV+CSFS+IKEY as a unit. CTV is already on inquisition, but if we want to stay in the spirit of the proposal, might it make more sense to have CSFS and IKEY be deployed as a single unit?

reardencode replied:
Yeah, IDK about what @ajtowns wants to do here for inquisition purposes. For core I'd probably want to do something with fewer activations and fewer flags.

rot13maxi commented on src/pubkey.cpp:
```diff
@@ -209,12 +209,12 @@ bool XOnlyPubKey::IsFullyValid() const
     secp256k1_xonly_pubkey pubkey;
     if (!secp256k1_xonly_pubkey_parse(secp256k1_context_static, &pubkey, m_keydata.data())) return false;
-    return secp256k1_schnorrsig_verify(secp256k1_context_static, sigbytes.data(), msg.begin(), 32, &pubkey);
+    return secp256k1_schnorrsig_verify(secp256k1_context_static, sigbytes.data(), msg.data(), msg.size(), &pubkey);
```
why did this change?

reardencode replied:
To allow data lengths other than 32. BIP340 explicitly allows it, but Taproot sighashes are all exactly 32-bytes so it wasn't needed previously.

ajtowns commented:
Did you want to continue with this on 25.x or skip straight to 27.x?

ecdsa commented:
@reardencode could you rebase this? I would love to test CSFS in combination with CAT

reardencode commented:
Rebased on 27.x.

ajtowns commented on test/functional/test_framework/script.py:
```diff
@@ -953,7 +953,7 @@ def taproot_construct(pubkey, scripts=None, *, keyver=None, treat_internal_as_in
-    # assume OP_CAT is activated in tests
-    if o == OP_CAT:
+    # assume new ops are activated in tests
+    if o == OP_CAT || o == OP_INTERNALKEY || o == OP_CHECKSIGFROMSTACK:
```
Wrong language...

instagibbs replied:
```suggestion
    if o in [OP_CAT, OP_INTERNALKEY, OP_CHECKSIGFROMSTACK]:
```

ajtowns commented on test/functional/test_framework/script.py:
```diff
@@ -953,7 +953,7 @@ def taproot_construct(pubkey, scripts=None, *, keyver=None, treat_internal_as_in
 
 def is_op_success(o):
-    # assume OP_CAT is activated in tests
-    if o == OP_CAT:
```
```
test/functional/test_framework/script.py:957:22: F821 undefined name 'OP_INTERNALKEY'
test/functional/test_framework/script.py:957:38: F821 undefined name 'OP_CHECKSIGFROMSTACK'
```

ajtowns commented:
This has been failing CI for over two months, so I'm closing this.

Per [twitter](https://x.com/4moonsettler/status/1860660928166944903), this is being pushed for activation in the near future, which I think will be a completely predictable trainwreck, so I personally won't be spending any more time reviewing this proposal. Beyond that, I'm skeptical that maintaining patches across two client (inquisition here and whatever the "activation client" is) will be a good idea: if signet and the activation client end up with incompatible implementations of some features, that seems like it would be pretty unhelpful.
